### ISSUE TITLE
visualizing the PPO probability distribution

### ISSUE THREAD

**Issue by jerabaul29** [reinforceio/tensorforce#481]
If I understand well, PPO produces a probability distribution for deciding the control. Out of this probability distribution, one either takes the max, or a random sample following the distribution.

- Is this probability distribution well a gaussian?

- Is there a way to visualize the internals of this probability distribution, for helping debug? For example, to visualize the mean (most likely value) and standard deviation (how unsure the network is). Otherwise, how can we implement this?

(this is what we discussed @UlysseREGLADE )

**Comment by AlexKuhnle** [comment]
The [type of distribution](https://github.com/reinforceio/tensorforce/tree/master/tensorforce/core/distributions) depends on the action type: bool --> Bernoulli, int --> Categorical, float --> Gaussian, float with min/max --> Beta.

Visualizing the values is quite straightforward: In case of a Gaussian, you can [here](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/core/distributions/gaussian.py#L66) add lines similar to e.g. [here](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/core/networks/layer.py#L395), and then add the summary labels to the agent config accordingly. I would suggest a summary label like `"distr-params"` or so, and then each distribution can store relevant values as part of the `parameterize(...)` function.

I could do this at some point, but it may be easiest/quickest if you implement it yourself. If you do, a PR would be welcome. :-)

**Comment by jerabaul29** [comment]
Ok, will look at it sometimes soon, may take a couple of days.

**Comment by AlexKuhnle** [comment]
No problem, let me know if there are questions. (y)

**Comment by UlysseREGLADE** [comment]
Hi everyone,

As far as I understand, you propose to add a tf.summary so that we can visualize the mean and std of the distribution within tensorboard. However, that's not exactly what we want to do.

Indeed, in our project, we have implemented an environment able to save and display in real time several variables during the training (reward, state, action...). This is achieved in the environment because some of these variable are not seen by the agent. We need to monitor them to ensure that everything is going the way it should during the training phase.

That's why, we where wondering if there was a way to get the probability distribution from the agent during the training in "real time", so that we can implement a graphical output as well as a data dumping to the disk.

I hope this is understandable and I'm clear enough...

**Comment by AlexKuhnle** [comment]
I see, in that case I can point you to the, unfortunately less-developed, feature of "named tensors", which can are retrieved as part of `model.act(...)` [here](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/models/model.py#L1377). Essentially you would need to add the same `get_named_tensor` functionality to the `Distribution` classes. An extension of this feature would of course equally be welcome as a PR, if you implement it yourself. :-)

**Comment by jerabaul29** [comment]
@UlysseREGLADE  how does this look? You were talking about both the 'clean' way to do this and a 'hacky' way by experimentally sampling the noisy PPO output. How does this look?

**Comment by UlysseREGLADE** [comment]
Hi there,
so I found a solution to my problem, it is a dirty one. Sorry, but this is not mature enough to deserve a pool request.

Here is how I did it:
-I dived into the tensorforce code to understand the architecture of the tensorflow graph.
-Our model has a single scalar output following a beta distribution, so my first step was to find the node of the graph corresponding to alpha and beta (I told you, it is dirty), as well as the placeholder for the state:

```Python

    with tf.Session() as sess:

        #Loading the meta graph
        new_saver = tf.train.import_meta_graph('saved_models/ppo_model.meta')

        #Loading the state of the graph
        new_saver.restore(sess, "saved_models/ppo_model")

        #Getting the default graph
        graph = tf.get_default_graph()

        alpha = tf.get_default_graph().get_tensor_by_name(u'ppo/actions-and-internals/beta/parameterize/alpha/apply/BiasAdd:0')
        beta = tf.get_default_graph().get_tensor_by_name(u'ppo/actions-and-internals/beta/parameterize/beta/apply/BiasAdd:0')
        state = tf.get_default_graph().get_tensor_by_name(u'ppo/state-state:0')
```

-However, I realized I needed to implement the activation function of alpha and beta myself, because the nodes above are the direct output of the linear layer:

```Python

        def active(x):
            x = np.clip(x, np.log(1e-6), -np.log(1e-6))
            x = np.log(np.exp(x) + 1.0) + 1.0
            return x

        def act(s):

            a, b = sess.run([alpha, beta], feed_dict={state:s})
            a, b = active(a), active(b)

            mean, var = ((b/max(a+b, 1e-6))-0.5)*2, (a*b/((a+b+1)*(a+b)**2))*2

            return mean, var
```

-Doing so I was able to use the graph to execute the model behavior, and extract the variance associated with the prediction of the trained model:

![control_var](https://user-images.githubusercontent.com/40145406/44984743-33869c00-af7e-11e8-9d13-517b091a0711.png)

**Comment by jerabaul29** [comment]
@AlexKuhnle do you think there is a way to make this approach of @UlysseREGLADE general by adding some helper function / thin wrappers on the agent? ;)

**Comment by AlexKuhnle** [comment]
Hmm, probably possible, yes. But easiest would be if the existing TensorBoard summary framework could be used. I assume that would also work? Retrieval of named tensors kind of exists, but ideally should be better designed in general.

**Comment by jerabaul29** [comment]
Yes, tensorboard could be very useful for that ;)

**Comment by AlexKuhnle** [comment]
The latest commit supports `'distribution'` as label in `summary_labels`, so hopefully should solve this issue.
